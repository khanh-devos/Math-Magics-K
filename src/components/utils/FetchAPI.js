import { useEffect, useState } from 'react';
import './utils.css';

const API = 'https://api.api-ninjas.com/v1/logo?name=';
const API_KEY = '8Dn4moNXRBFtBy/+FxeqaA==gcfYWmrBwNeX4ZuL';

export default function FetchAPI() {
  const [isLoading, setIsLoading] = useState(false);
  const [org, setOrg] = useState(null);
  const [logo, setLogo] = useState(null);
  const [message, setMessage] = useState(null);
  const handleSubmit = (e) => {
    e.preventDefault();
    const form = e.currentTarget;
    setOrg(form.org.value);
    form.reset();
  };

  useEffect(() => {
    const fetchAPI = async () => {
      try {
        const res = await fetch(API + org, {
          method: 'GET',
          headers: {
            'X-API-KEY': API_KEY,
          },
        });

        const [result] = await res.json();
        if (result) {
          setLogo(result.image);
          setMessage(null);
          setIsLoading(false);
        } else {
          setMessage(`There is no logo for this org "${org}", please add another one`);
          setTimeout(() => setMessage(null), 4000);
          setIsLoading(false);
        }
      } catch (err) {
        setIsLoading(false);
        setMessage(null);
        setMessage(err);
      }
    };

    if (org !== null) {
      setIsLoading(true);
      setLogo(null);
      fetchAPI();
    }
  }, [org]);

  if (isLoading) return (<div>Loading...</div>);

  return (
    <div className="fetch-api">
      <h5 className="api-header">When you input an organization name, it will show the logo of the organization</h5>

      <form onSubmit={handleSubmit}>
        <input name="org" className="input" placeholder="organization name" type="text" required />

        <input type="submit" className="items-shared" id="submit-btn" value="submit" />
      </form>

      {
        logo && <img src={logo} className="logo" alt="logo" width="200" />
      }

      {
        message && <div className="message">{message}</div>
      }

    </div>
  );
}
