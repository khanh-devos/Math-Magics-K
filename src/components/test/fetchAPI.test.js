import FetchAPI from '../utils/FetchAPI';
import { render, act } from '@testing-library/react';


describe('fetchAPI test', () => {
    // mock if dom container has been rendered
    it('should match the snapshot', () => {
        const { container } = render(<FetchAPI />);
        expect(container).toMatchSnapshot();
    });

    it('should render json()', async () => {
        jest.spyOn(global, 'fetch').mockImplementationOnce(() => Promise.resolve({
          json: () => Promise.resolve([{ quote: 'Khanh' }]),
        }));
    
        await act(async () => {
          const { asFragment } = render(<FetchAPI />);
          expect(asFragment).toMatchSnapshot();
        });
    });

    it('should render message', async () => {
        jest.spyOn(global, 'fetch').mockImplementationOnce(() => Promise.resolve({
          json: () => Promise.resolve([]),
        }));
    
        await act(async () => {
          const { asFragment } = render(<FetchAPI />);
          expect(asFragment).toMatchSnapshot();
        });
    });

    it('should render ERROR', async () => {
        jest.spyOn(global, 'fetch').mockImplementationOnce(() => "");
    
        await act(async () => {
          const { asFragment } = render(<FetchAPI />);
          expect(asFragment).toMatchSnapshot();
        });
    });
})