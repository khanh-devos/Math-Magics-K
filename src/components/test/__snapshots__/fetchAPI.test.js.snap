// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAPI test should match the snapshot 1`] = `
<div>
  <div
    class="fetch-api"
  >
    Loading...
  </div>
</div>
`;

exports[`fetchAPI test should render ERROR 1`] = `[Function]`;

exports[`fetchAPI test should render json() 1`] = `[Function]`;

exports[`fetchAPI test should render message 1`] = `[Function]`;
